src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,319 @@
+            _blockCount = 0;
+            _activeStart = 0;
+            _availableStart = 0;
+        }
+
+        public void Dispose()
+        {
+            _activeStart = 0;
+            _availableStart = 0;
+
+            if (_blocks is not null)
+            {
+                for (int i = 0; i < _blocks.Length; i++)
+                {
+                    if (_blocks[i] is not null)
+                    {
+                        ArrayPool<byte>.Shared.Return(_blocks[i]!);
+                        _blocks[i] = null;
+                    }
+                }
+
+                _blocks = null;
+            }
+        }
+
+        public MultiMemory ActiveMemory => _blocks is null ? MultiMemory.Empty : new MultiMemory(_blocks!, _activeStart, _availableStart - _activeStart);
+
+        public MultiMemory AvailableMemory => _blocks is null ? MultiMemory.Empty : new MultiMemory(_blocks!, _availableStart, _blockCount * BlockSize - _availableStart);
+
+        public void Discard(int byteCount)
+        {
+            Debug.Assert(byteCount >= 0);
+            Debug.Assert(byteCount <= ActiveMemory.Length, $"MultiArrayBuffer.Discard: Expected byteCount={byteCount} <= {ActiveMemory.Length}");
+
+            int oldStartBlock = _activeStart / BlockSize;
+            _activeStart += byteCount;
+            int newStartBlock = _activeStart / BlockSize;
+
+            while (oldStartBlock < newStartBlock)
+            {
+                Debug.Assert(_blocks is not null);
+                Debug.Assert(_blocks[oldStartBlock] is not null, $"Discard: oldStartBlock is null?? byteCount={byteCount}, _activeStart={_activeStart}, oldStartBlock={oldStartBlock}, newStartBlock={newStartBlock}");
+
+                ArrayPool<byte>.Shared.Return(_blocks[oldStartBlock]!);
+                _blocks[oldStartBlock] = null;
+
+                oldStartBlock++;
+            }
+
+            if (_activeStart == _availableStart)
+            {
+                // Small optimization to restart at the beginning of the current block, since we have no active bytes.
+                // Note, we don't try to release buffers in this case. Maybe we should. But if we did,
+                // we'd need handle the fact that there could be more than one here, if we previously grew the buffer enough for that.
+
+                _activeStart = newStartBlock * BlockSize;
+                _availableStart = newStartBlock * BlockSize;
+            }
+        }
+
+        public void Commit(int byteCount)
+        {
+            Debug.Assert(byteCount >= 0);
+            Debug.Assert(byteCount <= AvailableMemory.Length, $"MultiArrayBuffer.Commit: Expected byteCount={byteCount} <= {AvailableMemory.Length}");
+
+            _availableStart += byteCount;
+        }
+
+        // Ensure at least [byteCount] bytes to write to, up to the specified limit
+        public void TryEnsureAvailableSpaceUpToLimit(int byteCount, int limit)
+        {
+            if (ActiveMemory.Length >= limit)
```

Comment:
Might be worth having an ActiveMemoryLength property that just computes the needed length, rather than creating an ActiveMemory with a length only to then unwrap it to get back the length.  Same for AvailableMemory.

Reply:
Possibly. I've tried to err on the conservative side here, not adding stuff like that unless we determine we actually need it. The MultiMemory constructor is pretty cheap, so it seems like the JIT could inline it and optimize stuff away... that said, I don't know if it actually does this.

Looking at the places this is actually used, I think a more valuable change would probably be to avoid retrieving the ActiveMemory/AvailableMemory struct multiple times consecutively. Perhaps the JIT can optimize this too, but that doesn't seem super likely.

Thoughts?